include/RAJA/pattern/detail/reduce.hpp

```diff
@@ -0,0 +1,272 @@
+ *
+ ******************************************************************************
+ */
+template <typename T, template <typename, typename> class Combiner>
+class BaseReduceMin : public BaseReduce<T, RAJA::reduce::min, Combiner>
+{
+public:
+  using Base = BaseReduce<T, RAJA::reduce::min, Combiner>;
+  using Base::Base;
+
+  //! reducer function; updates the current instance's state
+  const BaseReduceMin &min(T rhs) const
+  {
+    this->combine(rhs);
+    return *this;
+  }
+};
+
+/*!
+ **************************************************************************
+ *
+ * \brief  MinLoc reducer class template for use in tbb execution.
+ *
+ **************************************************************************
+ */
+template <typename T, template <typename, typename> class Combiner>
+class BaseReduceMinLoc
+    : public BaseReduce<ValueLoc<T>, RAJA::reduce::min, Combiner>
+{
+public:
+  using Base = BaseReduce<ValueLoc<T>, RAJA::reduce::min, Combiner>;
+  using value_type = typename Base::value_type;
+  using Base::Base;
+
+  //! constructor requires a default value for the reducer
+  explicit BaseReduceMinLoc(T init_val, Index_type init_idx)
+      : Base(value_type(init_val, init_idx))
+  {
+  }
+
+  //! constructor requires a default value for the reducer
+  explicit BaseReduceMinLoc(value_type const & init_)
+      : Base(init_)
+  {
+  }
+
+  /// \brief reducer function; updates the current instance's state
+  const BaseReduceMinLoc &minloc(T rhs, Index_type loc) const
+  {
+    this->combine(value_type(rhs, loc));
+    return *this;
+  }
+
+  //! Get the calculated reduced value
+  Index_type getLoc() { return Base::get().getLoc(); }
+  //! Get the calculated reduced value
+  operator T() const { return Base::get(); }
+};
+
+/*!
+ **************************************************************************
+ *
+ * \brief  Max reducer class template for use in tbb execution.
+ *
+ **************************************************************************
+ */
+template <typename T, template <typename, typename> class Combiner>
+class BaseReduceMax : public BaseReduce<T, RAJA::reduce::max, Combiner>
+{
+public:
+  using Base = BaseReduce<T, RAJA::reduce::max, Combiner>;
+  using Base::Base;
+
+  //! reducer function; updates the current instance's state
+  const BaseReduceMax &max(T rhs) const
+  {
+    this->combine(rhs);
+    return *this;
+  }
+};
+
+/*!
+ **************************************************************************
+ *
+ * \brief  Sum reducer class template for use in tbb execution.
+ *
+ **************************************************************************
+ */
+template <typename T, template <typename, typename> class Combiner>
+class BaseReduceSum : public BaseReduce<T, RAJA::reduce::sum, Combiner>
+{
+public:
+  using Base = BaseReduce<T, RAJA::reduce::sum, Combiner>;
+  using Base::Base;
+
+  //! reducer function; updates the current instance's state
+  const BaseReduceSum &operator+=(T rhs) const
```

Comment:
reduction operators could also be CRTP inherited -- not sure if that's what we want.